docs/proposals/mutex-improvements.md

```diff
@@ -0,0 +1,96 @@
+# Proposal for Mutex/Semaphore improvements
+
+## Introduction
+
+The motivation for this is to improve the reliability of how locking works via mutexes and semaphores. Currently the implementation makes use of
+string formatting, this is not scalable (with respect to the size of developers and features).
+
+### How do mutexes currently work?
+
+Nearly all of the code regarding how mutexes work reside in `sync_manager.go`.
+Here is an example run of how locks are acquired and released. Some parts have been omitted for brevity, I recommend opening up the file and following through the
+examples below.
+
+`
+getHolderKey({"Namespace": "argo", Name: "example"}, "node") = "argo/example/node"
+
+-- MutexStatus After LockAcquired Call --
+items = ["argo", "example", "node"]
+holdingName = "node"
+ms.Holding = [MutexHolding{Mutex: lockKey, Holder: holdingName}]
+
+getResourceKey("argo", "example", "node") = "argo/example/node"
+`
+
+This works fine but let's examine another case where this breaks. This is the bug from issue <https://github.com/argoproj/argo-workflows/issues/8684>
+
+`
+getHolderKey({"Namespace": "argo", Name: "deadlock-test-sn8p5"}, "deadlock-test-sn8p5") = "argo/deadlock-test-sn8p5/deadlock-test-sn8p5"
+
+-- "MutexStatus" After LockAcquired Call --
+items = ["argo", "deadlock-test-sn8p5", "deadlock-test-sn8p5"]
+holdingName = "deadlock-test-sn8p5"
+ms.Holding = [MutexHolding{Mutex: lockKey, Holder: holdingName}]
+
+getResourceKey("argo", "deadlock-test-sn8p5", "deadlock-test-sn8p5") = "argo/deadlock-test-sn8p5"
+`
+
+### A criticism of the current approach
+
+The current approach is heavily dependent on string formatting. This makes it incredibly difficult to understand functionality when there are no comments or documentation
+outlining the logic used. On top of this, it makes the "Find References" functionality on LSP powered editors near useless. My understanding of the function `getResourceKey` is that
+it seems to construct the holder key that is generated via `getHolderKey` through using processed strings generated via `LockAcquired`.
+I strongly believe this needs to be refactored to be less dependent on string formatting. I strongly believe that the key obtaining process should be far more obvious, the disconnect between key acquisition and release is inherently flawed and will only lead to further bugs down the line.
+
+### How should they work?
+
+We should not be generating information after it has already been generated. It is extremely important to only maintain a single source of truth,
+generating this information (the holder key) produces sources of truth at every generation point. Ideally we should be storing this information somewhere
+as we will be needing it eventually when a release call is made.
+
+### Solutions
+
+#### Store the holder key directly in the MutexStatus/SemaphoreStatus structs
```

Comment:
How will these solutions form the holder keys?  Will it use string format?
Can you provide the sample yaml how it will look?